internal/backends/postgresql/query.go

```diff
@@ -0,0 +1,281 @@
+			}
+		default:
+			panic("Invalid error type: PathError expected")
+		}
+
+		switch v := rootVal.(type) {
+		case *types.Document:
+			iter := v.Iterator()
+			defer iter.Close()
+
+			// iterate through subdocument, as it may contain operators
+			for {
+				k, v, err := iter.Next()
+				if err != nil {
+					if errors.Is(err, iterator.ErrIteratorDone) {
+						break
+					}
+
+					return "", nil, lazyerrors.Error(err)
+				}
+
+				switch k {
+				case "$eq":
+					if f, a := filterEqual(p, rootKey, v); f != "" {
+						filters = append(filters, f)
+						args = append(args, a...)
+					}
+
+				case "$ne":
+					sql := `NOT ( ` +
+						// does document contain the key,
+						// it is necessary, as NOT won't work correctly if the key does not exist.
+						`_jsonb ? %[1]s AND ` +
+						// does the value under the key is equal to filter value
+						`_jsonb->%[1]s @> %[2]s AND ` +
+						// does the value type is equal to the filter's one
+						`_jsonb->'$s'->'p'->%[1]s->'t' = '"%[3]s"' )`
+
+					switch v := v.(type) {
+					case *types.Document, *types.Array, types.Binary,
+						types.NullType, types.Regex, types.Timestamp:
+						// type not supported for pushdown
+
+					case float64, bool, int32, int64:
+						filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next(), sjson.GetTypeOfValue(v)))
+						args = append(args, rootKey, v)
+
+					case string, types.ObjectID, time.Time:
+						filters = append(filters, fmt.Sprintf(sql, p.Next(), p.Next(), sjson.GetTypeOfValue(v)))
+						args = append(args, rootKey, string(must.NotFail(sjson.MarshalSingleValue(v))))
+
+					default:
+						panic(fmt.Sprintf("Unexpected type of value: %v", v))
+					}
+
+				default:
+					// $gt and $lt
+					// TODO https://github.com/FerretDB/FerretDB/issues/1875
+					continue
+				}
+			}
+
+		case *types.Array, types.Binary, types.NullType, types.Regex, types.Timestamp:
+			// type not supported for pushdown
+
+		case float64, string, types.ObjectID, bool, time.Time, int32, int64:
+			if f, a := filterEqual(p, rootKey, v); f != "" {
+				filters = append(filters, f)
+				args = append(args, a...)
+			}
+
+		default:
+			panic(fmt.Sprintf("Unexpected type of value: %v", v))
+		}
+	}
+
+	var filter string
+	if len(filters) > 0 {
+		filter = ` WHERE ` + strings.Join(filters, " AND ")
+	}
+
+	return filter, args, nil
+}
+
+// prepareOrderByClause adds ORDER BY clause with given sort document and returns the query and arguments.
+func prepareOrderByClause(p *metadata.Placeholder, sort *types.Document) (string, []any, error) {
+	iter := sort.Iterator()
+	defer iter.Close()
+
+	var key string
+	var order types.SortType
+
+	for {
+		k, v, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				break
+			}
+
+			return "", nil, lazyerrors.Error(err)
+		}
+
+		// Skip sorting if there are more than one sort parameters
+		if order != 0 {
+			return "", nil, nil
+		}
+
+		order, err = common.GetSortType(k, v)
```

Comment:
Ideally sort parameter is validated in handler level, because this returns `NewCommandErrorMsgWithArgument`. See also https://github.com/FerretDB/FerretDB/blob/main/internal/backends/backends.go#L34

Reply:
#3512